File: model/room.go
package model

import (
	"math/rand"
)

// TurnState represent the current state in a turn
type TurnState int

// TurnStatePlayer is set when the player can send answers
const TurnStatePlayer TurnState = 0

// TurnStateJudge is set when no one except the judge can vote
const TurnStateJudge TurnState = 1

// TurnStateShowWinner is set during a short period of time when the judge has sent its answer to be displayed to everyone
const TurnStateShowWinner TurnState = 2

// Room represents a room
type Room struct {
	RoomID           string
	Participants     []*User
	Displays         []*Display `json:"-"`
	ZenMode          bool
	Turn             int
	MaxTurn          int
	Started          bool
	CurrentBlackCard *Card
	DefaultRerollTimeout int

	TurnState        TurnState
	DefaultCountdown int
	CurrentCountdown int `json:"-"`

	SelectedCards []*Card `json:"-"`

	RemainingCards      []*Card `json:"-"`
	RemainingBlackCards []*Card `json:"-"`
	UsedCards           []*Card `json:"-"`
	UsedBlackCards      []*Card `json:"-"`

	JudgeSelection int
	Answers        []*Proposal
	WinningAnswer  *Proposal
	Winner         string

	AvailableDecks []*Deck
}

func (r *Room) HasEnoughCards() bool {
	amtCards := 0
	for _, c := range r.AvailableDecks {
		if c.IsSelected {
			amtCards = amtCards + c.AmtWhite
		}
	}
	return ((len(r.Participants)+1) * 7) <= amtCards
}

// IsReady returns whether the game can start
func (r *Room) IsReady() bool {
	return len(r.Participants) >= 3 && r.HasEnoughCards()
}

// PickCard picks a new white card and put it in the used backlog
func (r *Room) PickCard() *Card {
	amtRemaining := len(r.RemainingCards)

	// If there are no more cards, we refill
	if amtRemaining == 0 {
		// We put back all cards
		for _, c := range r.UsedCards {
			r.RemainingCards = append(r.RemainingCards, c)
		}

		r.UsedCards = []*Card{}

		// And we set back the remaining cards amount
		amtRemaining = len(r.RemainingCards)
	}

	cardPicked := rand.Intn(amtRemaining)
	card := r.RemainingCards[cardPicked]

	r.RemainingCards[cardPicked] = r.RemainingCards[len(r.RemainingCards)-1]
	r.RemainingCards = r.RemainingCards[:len(r.RemainingCards)-1]

	return card
}

// PickBlackCard picks a new black card and put it in the used backlog
func (r *Room) PickBlackCard() *Card {
	amtRemaining := len(r.RemainingBlackCards)

	// If there are no more cards, we refill
	if amtRemaining == 0 {
		for _, v := range r.UsedBlackCards {
			// We don't take the current card of course
			if v.ID != r.CurrentBlackCard.ID {
				r.RemainingBlackCards = append(r.RemainingBlackCards, v)
			}
		}

		r.UsedCards = []*Card{}
		amtRemaining = len(r.RemainingBlackCards)
	}

	cardPicked := rand.Intn(amtRemaining)
	card := r.RemainingBlackCards[cardPicked]

	r.RemainingBlackCards[cardPicked] = r.RemainingBlackCards[len(r.RemainingBlackCards)-1]
	r.RemainingBlackCards = r.RemainingBlackCards[:len(r.RemainingBlackCards)-1]

	r.UsedBlackCards = append(r.UsedBlackCards, card)

	return card
}


// Utils
func GetAllPlayersCardsID(r *Room) []int64 {
	cards := []int64{}
	for _, p := range r.Participants {
		for _, c := range p.Hand {
			if c != nil {
				cards = append(cards, c.ID)
			}
		}
	}
	return cards
}

func Contains(cards []int64, card int64) bool {
	for _, c := range cards {
		if c == card {
			return true
		}
	}
	return false
}

File: frontend/src/store/room.js
import Vue from 'vue';

export default {
    state: () => ({
        Answers: null,
        AvailableDecks: [],
        CurrentBlackCard: null,
        DefaultCountdown: 80,
        DefaultRerollTimeout: 4,
        JudgeSelection: -1,
        MaxTurn: 10,
        Participants: [],
        RoomID: '',
        Started: false,
        Turn: 0,
        TurnState: 0,
        Winner: '',
        WinningAnswer: null,
        ZenMode: false,
    }),
    mutations: {
        SET_GAMESTATE: (state, payload) => {
            Object.assign(state, payload.Room);
        },

        HAS_PLAYED: (state, payload) => {
            for (let i = 0; i < state.Participants.length; i++){
                if (state.Participants[i].Username === payload) {
                    Vue.set(state.Participants, i, { ...state.Participants[i], HasPlayed: true });
                }
            }
        },

        JUDGE_SELECTION: (state, payload) => {
            state.JudgeSelection = payload;
        },

        /**
         * Settings mutations
         */
        GOT_SETTINGS: (state, payload) => {
            state.MaxTurn = payload.MaxTurn;
            state.ZenMode = payload.ZenMode;
            state.DefaultCountdown = payload.DefaultCountdown;
            state.DefaultRerollTimeout = payload.DefaultRerollTimeout;

            for (let i = 0; i < state.AvailableDecks.length; i++) {
                let e = state.AvailableDecks[i];

                if (payload.SelectedDecks.includes(e.ID)) {
                    Vue.set(state.AvailableDecks[i], 'IsSelected', true);
                } else {
                    Vue.set(state.AvailableDecks[i], 'IsSelected', false);
                }
            }
        },
        updateTurns: (state, payload) => {
            state.MaxTurn = parseInt(payload) ?? 30;
        },
        updateZenMode: (state, payload) => {
            state.ZenMode = payload;
        },
        updateCountdown: (state, payload) => {
            state.DefaultCountdown = parseInt(payload) ?? 80;
        },
        updateRerollTimeout: (state, payload) => {
            state.DefaultRerollTimeout = parseInt(payload) ?? 6;
        },
        updateSelectedDecks: (state, payload) => {
            payload.ID = parseInt(payload.ID);

            for (let i = 0; i < state.AvailableDecks.length; i++) {
                if (state.AvailableDecks[i].ID === payload.ID) {
                    Vue.set(state.AvailableDecks, i, { ...state.AvailableDecks[i], IsSelected: payload.Selected });
                }
            }
        },
    },
    actions: {
        startGame({rootState}) {
            rootState.UI.WebSocket.send(JSON.stringify({ Command: 'START_GAME', Arguments: '{}' }));
        },
        sendSettings({rootState, state}) {
            let settings = {
                SelectedDecks: state.AvailableDecks.filter(e => e.IsSelected).map(e => e.ID),
                MaxTurn: state.MaxTurn,
                ZenMode: state.ZenMode,
                DefaultCountdown: state.DefaultCountdown,
                DefaultRerollTimeout: state.DefaultRerollTimeout,
            };

            rootState.UI.WebSocket.send(JSON.stringify({
                Command: 'SET_SETTINGS',
                Arguments: JSON.stringify(settings)
            }));
        },
    },
    getters: {
        IsReady: state => state.Participants.length >= 3,
        IsPlaying: state => state.ZenMode || state.Turn < state.MaxTurn,
        AmtCards: state => {
            let sumBlack = 0;
            let sumWhite = 0;

            let decks = state.AvailableDecks;

            if (!decks) {
                return { sumBlack: 0, sumWhite: 0 };
            }

            for (let i = 0; i < decks.length; i++) {
                if (decks[i].IsSelected) {
                    sumBlack += decks[i].AmtBlack;
                    sumWhite += decks[i].AmtWhite;
                }
            }

            return { sumBlack, sumWhite };
        },
        HasEnoughCards: (state, getters) => {
            return getters.AmtCards.sumWhite >= ((state.Participants.length+1) * 7);
        }
    }
};